docs/design-notes/raft-in-scylla.md

```diff
@@ -0,0 +1,207 @@
+which provides an ordered, at-most-once semantics of command
+application.
+
+This algorithm is using an own system table - `group0_history` - to
+store its state.
+
+The algorithm is adding a unique, monotonic, persistent state id
+to each command that is committed to Raft log of the group 0 state
+machine. This id is used
+to ensure that no two concurrent commands based on the same (and
+outdated) state of the client state machine are applied, and to
+verify that a command is committed to Raft log in a situation of
+uncertainty and retry the command if it failed.
+
+Here's how it works.
+
+Each change to the schema state machine is "signed", or augmented, with a
+state id - a unique monotonic identifier of the change. The state
+id is based on TIMEUUID. When the command is applied, the state id is
+persisted in `group0_history` table. In addition to the new state id,
+the previous state id is also saved in the command to check that
+the command is applied against the correct state of the client
+state machine.
+
+In order to perform a schema change, the client, which can be any Scylla
+node, performs the following steps:
+
+1. Takes a lock around its local client state machine, to protect against
+   concurrent operations of the state machine by different threads of
+   this node.
+1. Issues a raft read barrier. This ensures this local client
+   state machine has all the commands added to Group 0 on other
+   nodes, up to the time of the barrier.
+1. Reads `group0_history` table to find out the latest existing state id.
+1. Creates a new state id, which would identify the new command. It
+   is strictly greater than the previous state id.
+1. Reads the local state machine, validates and constructs a new command.
+   E.g. at this point, if we're creating a new keyspace, we may find
+   that the keyspace already exists and return an error to the user.
+   Otherwise we'll create a new mutation which adds the keyspace
+   to the system table.
+1. Creates a command to store in Raft log, which bundles together
+   the mutation of the client state machine, the previous and the new
+   state ids.
+1. When Raft commits the command to its log, each node will apply it
+   to its local schema state machine. At this point, verify that
+   the previous state id of the command matches the latest state
+   id of the local state machine, i.e. there was no race with another
+   Raft command from a different client, and if there is a match,
+   apply the command and append the new state id to `group0_history`,
+   otherwise turn the command into a no-op.
+1. The client which generated the command, can either wait for the
+   command to apply locally, or, if there was a timeout or error,
+   take a Raft read barrier, read the local state machien and find
+   out what happened with the command using `group0_history`. If the
+   state id is recorded in the history, the command is applied,
+   otherwise not, and will never be, so the whole procedure needs to
```

Comment:
That's not true. If there was a timeout, for example, the command may race with our read barrier and get ordered somewhere after the read barrier, so we cannot deduce that "it will never be" applied.

What we do instead is using the fact that the command is idempotent and just keep retrying it until it gets applied. Then we can check if it had an effect (i.e. if the state ID check was successful or not).